stdlib/public/core/ArrayBufferType.swift

```diff
@@ -157,28 +157,31 @@ extension _ArrayBufferType {
       newTailStart.moveInitializeBackwardFrom(oldTailStart, count: tailCount)
 
       // Assign over the original subRange
-      var i = newValues.startIndex
-      for j in subRange {
-        elements[j] = newValues[i]
-        i._successorInPlace()
+      var generator = newValues.generate()
+      for i in subRange { 
+        let element = generator.next()
+        _collectionAssert(element != nil)
```

Comment:
I haven't realized this was going to have a meaningful effect for -Onone. The original code was happy to use an out-of-bounds index if the collection was broken, but I think the consistent error is desirable. However, if it causes a slowdown, I agree it should be removed.

I'll take a look at the unoptimized benchmarks; if removing them makes an impact, I'll do so.